src/main/clojure/clara/rules/compiler.clj

```diff
@@ -1242,19 +1248,211 @@
 
 
 (sc/defn to-beta-graph :- schema/BetaGraph
-
   "Produces a description of the beta network."
-  [productions :- #{schema/Production}]
-
-  (let [id-counter (atom 0)
-        create-id-fn (fn [] (swap! id-counter inc))]
-
-    (reduce (fn [beta-graph production]
+  [productions :- #{schema/Production}
+   create-id-fn :- IFn]
+  (reduce (fn [beta-graph production]
               (binding [*compile-ctx* {:production production}]
                 (add-production production beta-graph create-id-fn)))
 
             empty-beta-graph
-            productions)))
+            productions))
+
+(sc/defn ^:private root-node? :- sc/Bool
+  "A helper function to determine if the node-id provided is a root node. A given node would be considered a root-node,
+   if its only backward edge was that of the artificial root node, node 0."
+  [backward-edges :- {sc/Int #{sc/Int}}
+   node-id :- sc/Int]
+  (= #{0} (get backward-edges node-id)))
+
+(sc/defn extract-exprs :- schema/NodeExprLookup
+  "Walks the Alpha and Beta graphs and extracts the expressions that will be used in the construction of the final network.
+   The extracted expressions are stored by their key, [<node-id> <field-key>], this allows for the function to be retrieved
+   after it has been compiled.
+
+   Note: The keys of the map returned carry the metadata that can be used during evaluation. This metadata will contain,
+   if available, the compile context, file and ns. This metadata is not stored on the expression itself because it will
+   contain forms that will fail to eval."
+  [beta-graph :- schema/BetaGraph
+   alpha-graph :- [schema/AlphaNode]]
+  (let [backward-edges (:backward-edges beta-graph)
+        handle-expr (fn [id->expr s-expr id expr-key metadata]
+                      (assoc id->expr
+                        ;; Letting the key carry all of the metadata, this makes the batch evaluation of the expressions
+                        ;; easier because the metadata contains unevalable forms.
+                        (with-meta [id expr-key] (assoc metadata expr-key s-expr))
```

Comment:
This pattern of having the metadata carry fundamental semantic meaning of the data (without which durability would be broken) that isn't reflected in the NodeExprLookup schema seems like a suboptimal practice.  I'd like to at least have the metadata expectations reflected in the schema.  Alternatively you could have the key map to a tuple, map, etc. containing both the values, but my primary concern here is that as someone reading the code, I'd expect the schema to be a complete description of the data, or at least the mandatory parts of the data.  This seems to break that expectation.

```
because the metadata contains unevalable forms.
```
Are you referring to the issues discussed in https://github.com/cerner/clara-rules/pull/243 ?  If so could you reference it in the comment?